### PR TITLE
Fix gate power for arbitrary gates

### PR DESCRIPTION
**Pull request by ajavadia**
gate.power() was relying on gate.to_matrix(), which does not always exist for arbitrary gates. However we can get the matrix using `Operator(gate).data`.

### CODE REVIEW

File: qiskit/circuit/gate.py [hunk 1]
```diff
@@ -58,9 +58,10 @@ def power(self, exponent):
         Raises:
             CircuitError: If Gate is not unitary
         """
+        from qiskit.quantum_info.operators import Operator
         from qiskit.extensions.unitary import UnitaryGate  # pylint: disable=cyclic-import
         # Should be diagonalized because it's a unitary.
-        decomposition, unitary = schur(self.to_matrix(), output='complex')
+        decomposition, unitary = schur(Operator(self).data, output='complex')
```
Comment (ewinston) [inline qiskit/circuit/gate.py]: Could we just make this the default implementation of `to_matrix` for the Gate class?

Comment (ajavadia) [reply qiskit/circuit/gate.py]: potentially - but the reason it has not been done yet is because of global phase ambiguities in the definitions.

**Comment by 1ucian0** [comment]
Do you mind adding a test case?

File: test/python/circuit/test_gate_power.py [hunk 2]
```diff
@@ -110,7 +108,23 @@ def test_standard_sqrt(self):
...
+
+    def test_composite_sqrt(self):
+        """Test composite Gate.power(1/2) method.
+        """
+        circ = QuantumCircuit(1, name='my_gate')
+        circ.rz(0.1, 0)
+        circ.rx(0.2, 0)
+        gate = circ.to_gate()
+
+        expected = array([[0.99874948+6.25390559e-05j,  0.00374609-4.98542083e-02j],
```
Comment (kdk) [inline test/python/circuit/test_gate_power.py]:
```suggestion
        expected = array([[0.99874948+6.25390559e-05j, 0.00374609-4.98542083e-02j],
```

File: test/python/circuit/test_gate_power.py [hunk 3]
```diff
@@ -110,7 +108,23 @@ def test_standard_sqrt(self):
...
+                          [-0.00124974-4.99791301e-02j,  0.99750443+4.98542083e-02j]])
```
Comment (kdk) [inline test/python/circuit/test_gate_power.py]:
```suggestion
                          [-0.00124974-4.99791301e-02j, 0.99750443+4.98542083e-02j]])
```

File: qiskit/converters/circuit_to_gate.py [hunk 4]
```diff
@@ -42,6 +42,10 @@ def circuit_to_gate(circuit, parameter_map=None):
         input circuit. Upon decomposition, this gate will
         yield the components comprising the original circuit.
     """
+    if len(circuit.clbits) != 0:
```
Comment (kdk) [inline qiskit/converters/circuit_to_gate.py]:
```suggestion
    if circuit.clbits:
```

File: qiskit/circuit/gate.py [hunk 5]
```diff
@@ -58,9 +58,10 @@ def power(self, exponent):
         Raises:
             CircuitError: If Gate is not unitary
         """
+        from qiskit.quantum_info.operators import Operator
```
Comment (kdk) [inline qiskit/circuit/gate.py]:
```suggestion
        from qiskit.quantum_info.operators import Operator. # pylint: disable=cyclic-import
```

File: test/python/circuit/test_gate_power.py [hunk 6]
```diff
@@ -57,7 +55,7 @@ def test_sgate_float(self, n):
         expected = self.results[n]
         self.assertEqual(result.label, 's^%s' % n)
         self.assertIsInstance(result, UnitaryGate)
-        assert_array_almost_equal(result.to_matrix(), expected)
+        self.assertTrue(Operator(result), Operator(expected))
```
Comment (kdk) [inline test/python/circuit/test_gate_power.py]:
```suggestion
        self.assertEqual(Operator(result), Operator(expected))
```

File: test/python/circuit/test_gate_power.py [hunk 7]
```diff
@@ -80,7 +78,7 @@ def test_cx_int(self, n):
 
         self.assertEqual(result.label, 'cx^' + str(n))
         self.assertIsInstance(result, UnitaryGate)
-        assert_array_almost_equal(result.to_matrix(), self.results[n])
+        self.assertTrue(Operator(result), Operator(self.results[n]))
```
Comment (kdk) [inline test/python/circuit/test_gate_power.py]:
```suggestion
        self.assertEqual(Operator(result), Operator(self.results[n]))
```

File: test/python/circuit/test_gate_power.py [hunk 8]
```diff
@@ -125,8 +141,8 @@ def test_direct_root(self, degree):
         self.assertEqual(result.label, 's^' + str(1 / degree))
         self.assertEqual(len(result.definition), 1)
         self.assertIsInstance(result, Gate)
-        assert_allclose(matrix_power(result.definition[0][0].to_matrix(), degree),
-                        SGate().to_matrix())
+        self.assertTrue(Operator(result).power(degree),
```
Comment (kdk) [inline test/python/circuit/test_gate_power.py]:
```suggestion
        self.assertEqual(Operator(result).power(degree),
```

**Comment by ajavadia** [comment]
@kdk thanks for catching those.

File: test/python/circuit/test_gate_power.py [hunk 9]
```diff
@@ -125,8 +141,8 @@ def test_direct_root(self, degree):
         self.assertEqual(result.label, 's^' + str(1 / degree))
         self.assertEqual(len(result.definition), 1)
         self.assertIsInstance(result, Gate)
-        assert_allclose(matrix_power(result.definition[0][0].to_matrix(), degree),
-                        SGate().to_matrix())
+        self.assertEqual(Operator(result).power(degree),
+                        Operator(SGate()))
```
Comment (kdk) [inline test/python/circuit/test_gate_power.py]:
```suggestion
                         Operator(SGate()))
```

**Pull request merged**